File: python/03-sliding-window/max_sum_subarray.py
"""
Problem: Maximum Sum Subarray of Size K
Pattern: Sliding Window (fixed size)
Time: O(n) | Space: O(1)

Input:  nums = [2, 1, 5, 1, 3, 2], k = 3
Output: 9  (subarray [5, 1, 3])
"""

def max_sum_subarray(nums, k):
    window_sum = sum(nums[:k])
    max_sum = window_sum
    for i in range(k, len(nums)):
        window_sum += nums[i] - nums[i - k]
        max_sum = max(max_sum, window_sum)
    return max_sum


# Alternative: brute force -- recompute the sum of every window from scratch.
# Time: O(n*k) instead of O(n). Shows the naive first-attempt vs the
# optimized sliding-window version above.
def max_sum_subarray_brute_force(nums, k):
    max_sum = float('-inf')
    for i in range(len(nums) - k + 1):
        window_sum = sum(nums[i:i + k])
        max_sum = max(max_sum, window_sum)
    return max_sum


if __name__ == "__main__":
    print(max_sum_subarray([2, 1, 5, 1, 3, 2], 3))  # 9
    print(max_sum_subarray_brute_force([2, 1, 5, 1, 3, 2], 3))  # 9

File: javascript/03-sliding-window/max_sum_subarray.js
/**
 * Problem: Maximum Sum Subarray of Size K
 * Pattern: Sliding Window (fixed size)
 * Time: O(n) | Space: O(1)
 *
 * Input:  nums = [2, 1, 5, 1, 3, 2], k = 3
 * Output: 9  (subarray [5, 1, 3])
 */

function maxSumSubarray(nums, k) {
    let windowSum = 0;
    for (let i = 0; i < k; i++) windowSum += nums[i];
    let maxSum = windowSum;

    for (let i = k; i < nums.length; i++) {
        windowSum += nums[i] - nums[i - k];
        maxSum = Math.max(maxSum, windowSum);
    }
    return maxSum;
}

// Alternative: brute force -- recompute the sum of every window from
// scratch. Time: O(n*k) instead of O(n) -- included to make the sliding
// window's benefit concrete (this is what you'd naively write first,
// then optimize into the version above).
function maxSumSubarrayBruteForce(nums, k) {
    let maxSum = -Infinity;
    for (let i = 0; i <= nums.length - k; i++) {
        let windowSum = 0;
        for (let j = i; j < i + k; j++) {
            windowSum += nums[j];
        }
        maxSum = Math.max(maxSum, windowSum);
    }
    return maxSum;
}

console.log(maxSumSubarray([2, 1, 5, 1, 3, 2], 3)); // 9

module.exports = { maxSumSubarray, alternative: maxSumSubarrayBruteForce };
